ibc/client/submodule.go

```diff
@@ -0,0 +1,225 @@
+		c.logger.Error().Err(err).Str("identifier", identifier).
+			Msg("failed to initialize client")
+		return "", err
+	}
+
+	c.logger.Info().Str("identifier", identifier).
+		Str("height", clientState.GetLatestHeight().ToString()).
+		Msg("client created at height")
+
+	// Emit the create client event to the event logger
+	if err := c.emitCreateClientEvent(identifier, clientState); err != nil {
+		c.logger.Error().Err(err).Str("identifier", identifier).
+			Msg("failed to emit client created event")
+		return "", err
+	}
+
+	return identifier, nil
+}
+
+// UpdateClient updates an existing client with the given identifier using the
+// ClientMessage provided
+func (c *clientManager) UpdateClient(
+	identifier string, clientMessage modules.ClientMessage,
+) error {
+	// Get the client state
+	clientState, err := c.GetClientState(identifier)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve the client store prefixed with the client identifier
+	prefixed := path.ApplyPrefix(core_types.CommitmentPrefix(path.KeyClientStorePrefix), identifier)
+	clientStore, err := c.GetBus().GetIBCHost().GetProvableStore(string(prefixed))
+	if err != nil {
+		return err
+	}
+
+	// Check the state is active
+	if clientState.Status(clientStore) != modules.ActiveStatus {
+		return core_types.ErrIBCClientNotActive()
+	}
+
+	// Verify the client message
+	if err := clientState.VerifyClientMessage(clientStore, clientMessage); err != nil {
+		return err
+	}
+
+	// Check for misbehaviour on the source chain
+	misbehaved := clientState.CheckForMisbehaviour(clientStore, clientMessage)
+	if misbehaved {
+		if err := clientState.UpdateStateOnMisbehaviour(clientStore, clientMessage); err != nil {
+			c.logger.Error().Err(err).Str("identifier", identifier).
+				Msg("failed to freeze client for misbehaviour")
+			return err
+		}
+		c.logger.Info().Str("identifier", identifier).
+			Msg("client frozen for misbehaviour")
+
+		// emit the submit misbehaviour event to the event logger
+		if err := c.emitSubmitMisbehaviourEvent(identifier, clientState); err != nil {
+			c.logger.Error().Err(err).Str("identifier", identifier).
+				Msg("failed to emit client submit misbehaviour event")
+			return err
+		}
+		return nil
+	}
+
+	// Update the client
+	consensusHeight, err := clientState.UpdateState(clientStore, clientMessage)
+	if err != nil {
+		c.logger.Error().Err(err).Str("identifier", identifier).
+			Str("height", consensusHeight.ToString()).
+			Msg("failed to update client state")
+		return err
+	}
+	c.logger.Info().Str("identifier", identifier).
+		Str("height", consensusHeight.ToString()).
+		Msg("client state updated")
+
+	// emit the update client event to the event logger
+	if err := c.emitUpdateClientEvent(identifier, clientState.ClientType(), consensusHeight, clientMessage); err != nil {
+		c.logger.Error().Err(err).Str("identifier", identifier).
+			Msg("failed to emit client update event")
+		return err
+	}
+
+	return nil
+}
```

Comment:
tl;dr Create a ticket to scope (not implement) E2E feature tests

---

Lots of E2E business logic happening here that I can't provide much input on.

Consider writing some additional (non-functional) `.feature` files for E2E tests. 

It can totally be a separate ticket since even scoping them (not implementing them) is a nontrivial amount of work.

I think there's an opportunity to move this entire diff into an LLM (maybe claude since the context window is larger) and having it output some E2E tests.